Troubleshooting section major update (#705)
graknlabs/docs: typedb-src/modules/ROOT/pages/development/best.adoc

  * `relation`
  * `attribute`

> An *attribute type* (a subtype of the `attribute` root type) is a single concept with a fixed value in a given domain.

Reviewer: I'm not sure what "concept" means here. Is this terminology explained somewhere, or should we consider a different vocabulary?

Author: It's defined in https://typedb.com/docs/typedb/2.x/fundamentals/patterns 
Do you think it is not clear enough? This Troubleshooting section is quite advanced, so it requires a lot of context, that can be obtained from other pages of our documentation, that supposed to be read berfore this one.

Reviewer: I agree. Will the right links be obvious in the sidebar?

That attribute is a property of the domain rather than of its owners and so is not uniquely defined by any concepts
that might own it, nor does it require any owners to exist: a name of value "Bob" can be owned by either no one, one
particular person, or multiple people.

An *entity type* (a subtype of the `entity` root type) is a single concept with an independent existence. It might
practically require other entities to exist, such as a car that cannot exist without its parts but can be
conceptualized without reference to those other entities: a car can be imagined without considering its parts.

A *relation type* (a subtype of the `relation` root type) is a single concept with an existence that depends on at
least one other concept. It cannot be conceptualized without those concepts: it is impossible to imagine a marriage
without considering its spouses.

=== Tips for beginners

* The best way to start modeling a database schema is to model entity types first and then proceed with connecting
them with relation types and assign ownership of attribute types to them later.

> * The schema of a DB defines a vocabulary for queries to that particular DB. The closer the schema resembles the real-life

Reviewer: ```suggestion
* The schema defines the vocabulary for queries to a particular database. The closer the schema resembles the real-life
```

> data, the simpler it is to read queries for people familiar with the terminology. Consider putting extra effort into

Reviewer: ```suggestion
data, the simpler it is for people familiar with the terminology to read and understand the queries. Consider putting extra effort into
```

> the following aspects of designing a DB schema: accurate type labels/names, an intuitive hierarchy of subtypes, and

Reviewer: ```suggestion
the following aspects of designing your schema: an intuitive hierarchy of subtypes, meaningfully designed relations, and accurate type labels and names.
```

Author: We are not using you/your pronouns in our docs as per the existing style guide. That might change soon, as we are developing a new style guide for the docs with Caralyn, but to maintain consistency, we should not use it now.

Author: ```suggestion
the following aspects of designing a DB schema: an intuitive hierarchy of subtypes, meaningfully designed relations, and accurate type labels and names.
```

> designing relations in the most meaningful way.

Reviewer: ```suggestion
```

The above query returns full-names (`$n`) owned by `$p` of the `person` type.

> Can we expect to have a full name for every person instance in the results? *No*.

Reviewer: ```suggestion
Can we expect every person instance in the results to have a full name? *No*.
```

Author: If I understand correctly, that changes the meaning of the phrase, making it ambiguous. We probably need a new variant to maintain the correct meaning and unambiguousness. 
```suggestion
Can we expect the response to have the full-names of every person instance? *No*.
```

. A person can have more than one attribute of type `full-name`. Every instance of attribute will get to the results.
. A person can have no attributes of type `full-name`. In that case, the person will not be represented by the variable
`$p`. That will person will not be accounted for.
. Finally, different people can have the same full names. In TypeDB database that means different instances of `person`
type can own the same instance of the `full-name` attribute type. By filtering results to get only full-names we will
receive a deduplicated list of full-names, owned by any `person` instance, because it's just a list of all attributes
owned by the `$p`. We can't get the same answer more than once in a set.

> To get complete information about all full-names of every person, we need to modify the query as follows:

Reviewer: This whole section is inconsistent between "full-names" and "full names" (no hyphen), even when it's not the variable `full-name`.

Author: There were only two cases of it without a hyphen, and I've fixed one of them.
I've tried to make a difference between the instances of a type (with a hyphen, how we see it in a variable) and the semantics behind it (without). 
Do you think we shouldn't do that?

their owner. If a person has two `full-name` attributes, then the resulting response will contain two pairs with the
same `person` object but different `attributes`.

We can further improve the output by xref:typedb:ROOT:fundamentals/queries.adoc#_modifiers[grouping] the results by
`person` and/or applying xref:typedb:ROOT:fundamentals/queries.adoc#_modifiers[aggregation] to count the number of
results.

=== The uniqueness of attributes

TypeDB considers attributes to be first-class features and stores only unique values.

Any instance of any attribute type is unique. But ownership of that instance of an attribute type can be assigned any
number of times -- from zero to many. If multiple instances of some different types all have ownership over an attribute
with the same type and value, then they all have ownership of the same single instance.

Any attempt to insert a new instance of an attribute type with an existing value will not create a new instance.
Suppose we insert ownership of an attribute type instance that was not inserted before.
In that case, we achieve that by implicitly creating the attribute instance we assign ownership.

> This contrasts with an SQL model where "attributes" are columns in a table, and values can be repeated in each row.

Reviewer: "Contrast" is a transitive verb.

Author: ```suggestion
This is in contrast to the SQL model where "attributes" are columns in a table, and values can be repeated in each row.
```
